ghga_datasteward_kit/s3_upload/entrypoint.py

```diff
@@ -0,0 +1,230 @@
+    objectstorage,
+)
+from ghga_datasteward_kit.utils import load_config_yaml, read_token
+
+
+async def shared_path(input_path: Path, alias: str, config: LegacyConfig):
+    """
+    Functionality shared between legacy and non-legacy upload path
+    """
+    if not input_path.exists():
+        msg = f"No such file: {input_path.resolve()}"
+        handle_superficial_error(msg=msg)
+
+    if input_path.is_dir():
+        msg = f"File location points to a directory: {input_path.resolve()}"
+        handle_superficial_error(msg=msg)
+
+    check_output_path(config.output_dir / f"{alias}.json")
+
+    file_size = input_path.stat().st_size
+    check_adjust_part_size(config=config, file_size=file_size)
+
+    # set retry policy
+    HttpxClientState.configure(5)
+
+    uploader = ChunkedUploader(
+        input_path=input_path,
+        alias=alias,
+        config=config,
+        unencrypted_file_size=file_size,
+    )
+    await uploader.encrypt_and_upload()
+
+    downloader = ChunkedDownloader(
+        config=config,
+        file_id=uploader.file_id,
+        encrypted_file_size=uploader.encryptor.encrypted_file_size,
+        file_secret=uploader.encryptor.file_secret,
+        part_size=config.part_size,
+        target_checksums=uploader.encryptor.checksums,
+    )
+    await downloader.download()
+
+    return uploader, file_size
+
+
+async def exchange_secret_for_id(
+    *, file_id: str, alias: str, secret: bytes, token: str, config: Config
+) -> str:
+    """
+    Call file ingest service to store the file secret and obtain a secret ID by which
+    it can be retrieved.
+
+    If storing the secret fails, the uploaded file is deleted from object storage and
+    a ValueError is raised containing the file alias and response status code.
+    """
+
+    file_secret = base64.b64encode(secret).decode("utf-8")
+    payload = encrypt(data=file_secret, key=config.secret_ingest_pubkey)
+    encrypted_secret = models.EncryptedPayload(payload=payload)
+
+    with httpx_client() as client:
+        headers = {"Authorization": f"Bearer {token}"}
+        response = client.post(
+            url=config.secret_ingest_url,
+            json=encrypted_secret.dict(),
+            headers=headers,
+            timeout=60,
+        )
+
+        if response.status_code != 200:
+            object_storage = objectstorage(config=config)
+            await object_storage.delete_object(
+                bucket_id=config.bucket_id, object_id=file_id
+            )
+            raise ValueError(
+                f"Failed to deposit secret for {alias} with response code"
+                + f" {response.status_code}. Removed uploaded file."
+            )
+        return response.json()["secret_id"]
+
+
+async def async_main(input_path: Path, alias: str, config: Config, token: str):
+    """
+    Run encryption, upload and validation.
+    Prints metadata to <alias>.json in the specified output directory
+    """
+
+    uploader, file_size = await shared_path(
```

Comment:
Maybe in general, a context manager at this level that just cleans up when something goes wrong would be an option. What do you think?

Reply:
Currently there's a context manager for the multipart upload that also deals with cleanup in that context.
Moving the cleanup functionality into a dedicated context manager on a higher level would make sense, but the problem here is getting the upload ID to the correct place, which is only generated on initializing the multipart upload.

Reply:
Probably could do that using a an attribute on a custom exception, I'll see if I can come up with something simple

Reply:
Sounds good. But no worries if it becomes to difficult.